### ISSUE TITLE
PostgreSQL tables are not listed in Distiller

### ISSUE THREAD

**Issue by niroshansb** [mapmint/mapmint#11]
Trying to create a SQL views using tables without geometry column 

Step 1: Successfully connected to PostgreSQL db on localhost 
Step 2: Selecting the database in Datastore does not list existing tables in the db 
  
Running mapmint on OSGeoLive 11.0

**Comment by gfenoy** [comment]
From the Distiller module, you should be able to create a view by using the Process button on an existing layer in your database. So the only requirement is to have at least one geographic table already available in your database before you can proceed.

In my case, I have geographic data in the mmdb datastore that corresponds to a database connection.

First I open the datastore from the Distiller then I access the table I want to use to create a view by pressing the "Access database" button from the corresponding datastore displayed on the main panel. From here I can see the fields definitions and the values available in the table.

Then, I open the process panel by pressing the process button on a datasource available in the mmdb datastore. From here, I set the value for the target datastore parameter to the current one (so mmdb).  Then I check the SQL checkbox and use the followig SQL query:

`CREATE VIEW examples.new_examples_demo_liens AS (SELECT id, placette, camp ST_SetSRID(ST_MakePoint(x,y),32628) FROM your_schema.your_table);`

You can see bellow the different panels used to proceed:
 
<img width="1372" alt="capture d ecran 2017-12-06 10 02 50" src="https://user-images.githubusercontent.com/1606022/33653575-c847b89c-da6d-11e7-98c9-7ecff1668bad.png">

From here, you can then press the "Run convertion" button to proceed. Once the conversion is done, you simply have to refresh your datastore to see the new geographic tables / views listed. In my case, if I display the table, by pressing the "Toggle table" button of the datastore corresponding to my view, I get the following result.

<img width="1334" alt="capture d ecran 2017-12-06 10 03 10" src="https://user-images.githubusercontent.com/1606022/33653712-36dc589e-da6e-11e7-8856-87e7636212f0.png">

I hope you can confirm that its works also for you.

**Comment by niroshansb** [comment]
In my case, When I try "Access database"  it returns the following error 


Unable to run the Service. The message returned back by the Service was the following: Unable to execute SELECT DISTINCT on ("Pos") "Pos"-1 as "Pos","Field","Type","Key", "Ref", "RefCol", "RefCols",array_upper("RefCols",1) from (SELECT * FROM (SELECT attnum AS "Pos", attname AS "Field",CASE WHEN atttypmod >0 THEN b.typname || '(' || atttypmod-4 || ')' ELSE b.typname END AS "Type" FROM pg_catalog.pg_attribute a, pg_catalog.pg_type b WHERE a.atttypid=b.oid AND a.attrelid = (SELECT pg_class.oid FROM pg_class, pg_namespace WHERE relname='BIMARI_CORE' AND pg_namespace.oid=relnamespace AND nspname='odk_prod') AND a.attnum > 0 AND NOT a.attisdropped ORDER BY attnum) a LEFT JOIN (SELECT conkey,c.conname AS constraint_name, CASE c.contype WHEN 'c' THEN 'CHECK' WHEN 'f' THEN 'FOR' WHEN 'p' THEN 'PRI' WHEN 'u' THEN 'UNIQUE' END AS "Key", t3.nspname||'.'||t2.relname AS "Ref", (SELECT attname from pg_catalog.pg_attribute WHERE attrelid=c.confrelid AND confkey[1] = attnum) AS "RefCol" FROM pg_constraint c LEFT JOIN pg_class t ON c.conrelid = t.oid LEFT JOIN pg_class t2 ON c.confrelid = t2.oid LEFT JOIN pg_namespace t3 ON t2.relnamespace=t3.oid WHERE t.relname = 'BIMARI_CORE' and t.relnamespace=(select oid from pg_namespace where nspname='odk_prod')) b ON get_nb_of(conkey,"Pos")>0 LEFT JOIN (SELECT DISTINCT ON (at2.attnum) c.*, at2.attnum AS "myid", ARRAY(SELECT attname AS "RefCol" FROM pg_constraint AS c, pg_catalog.pg_attribute, pg_class t, pg_class t2 WHERE c.conrelid = t.oid AND c.confrelid = t2.oid AND t.relname = 'BIMARI_CORE' AND attrelid=confrelid AND get_nb_of(confkey,attnum) > 0 and t.relnamespace=(select oid from pg_namespace where nspname='odk_prod')) AS "RefCols", at2.attnum, at2.attname AS atn, get_index_of(conkey,at2.attnum) AS "RealOrigColNum", at1.attnum, at1.attname, get_index_of(confkey,at1.attnum) AS "RealRefColNum", t.relname as orig, t2.relname as ref FROM pg_constraint AS c, pg_catalog.pg_attribute AS at1, pg_catalog.pg_attribute AS at2, pg_class t, pg_class t2 WHERE c.conrelid = t.oid AND c.confrelid = t2.oid AND t.relname = 'BIMARI_CORE' AND at1.attrelid=confrelid AND get_nb_of(conkey,at2.attnum) > 0 AND get_nb_of(confkey,at1.attnum) > 0 AND t.relname='BIMARI_CORE' AND at2.attrelid=t.oid and t.relnamespace=(select oid from pg_namespace where nspname='odk_prod')) AS foreigns ON foreigns.myid=a."Pos" order by "Key"='PRI' or "Key"='FOR' desc) As f due to: function get_nb_of(smallint[], smallint) does not exist LINE 1: ...from pg_namespace where nspname='odk_prod')) b ON get_nb_of(... ^ HINT: No function matches the given name and argument types. You might need to add explicit type casts.

After that when I am trying to create a view by using the following code 

CREATE VIEW odk_prod.newlayer AS (SELECT "_URI", "LOCATION_LNG", "LOCATION_LAT" ST_SetSRID(ST_MakePoint("LOCATION_LNG", "LOCATION_LAT"),4326) FROM odk_prod."BIMARI_CORE");

It did not return any error and while I am refreshing also It did not return any error. 
And says 

Dir cleaned up
please wait while load 3 datasources from odk datastore. 

Thanks!

**Comment by gfenoy** [comment]
Thanks for following up on this issue @niroshansb.

It sounds strange that your database doesn't contain the `get_nb_of` definition cause it is defined from [here](https://github.com/mapmint/mapmint/blob/master/template/sql/mmdb.sql#L20). To solve this first issue, I invite you to load the definition of the function from the file you should have imported initially.


Your SQL query is wrong, you forgot to add a comma between your last field ("LOCATION_LAT" ) and the function call to create the geometry field.

I guess that this version of the SQL request should work properly:

`CREATE VIEW odk_prod.newlayer AS (SELECT "_URI" as uri, "LOCATION_LNG" as long, "LOCATION_LAT" as lat, ST_SetSRID(ST_MakePoint("LOCATION_LNG", "LOCATION_LAT"),4326) as wkb_geometry FROM odk_prod."BIMARI_CORE");`

Note that I have used lower case definition for the field name to avoid any issue that may be faced in case you use names starting by "_" for instance.

I hope this helps.

**Comment by niroshansb** [comment]
We have created get_nb_of function in our database. Finally we were able to open "Access database" option without any error. But it does not show the data in the table. It shows a message as "Processing.." and remains. 

![bimari](https://user-images.githubusercontent.com/10414489/33757401-4b26a590-dc3d-11e7-9971-3b8cd3a22423.png)
Since you mentioned you have tested in the data in mmdb.
We have copied and created a table in the public schema inside the mmdb as bimari. So it also gives the same result as shown in the above image. 

We have tried the SQL command you corrected. It did not create a view. 
Therefor I am attaching the table with the content for your reference. 

[data.zip](https://github.com/mapmint/mapmint/files/1542005/data.zip)

Note: These tables and data are created automatically when we uploading forms to the database, it is creating definitions, tables and db s with uppercase.

**Comment by gfenoy** [comment]
I have imported your file in the database corresponding to my mmdb datastore.

I am not able to view the table content as in your setup. This issue shall be reported in another ticket. Indeed, MapMint never had supported postgresql table that use upper cases or special char as the first field  char for its name. But it can be handled but it is another issue...

Then I have run the following request from the process panel:

`CREATE VIEW public.odk_niroshan_newlayer AS (SELECT "_URI" as uri, "LOCATION_LNG" as long, "LOCATION_LAT" as lat, ST_SetSRID(ST_MakePoint("LOCATION_LNG", "LOCATION_LAT"),4326) as wkb_geometry FROM public.bimari);`

It has successfully created the requested view accessible from the Distilller after refreshing the corresponding datastore, as it is shown bellow.

<img width="1465" alt="capture d ecran 2017-12-08 10 14 38" src="https://user-images.githubusercontent.com/1606022/33763784-195fa1cc-dc12-11e7-9fb6-bc6c0d4e87ac.png">
